### PR TITLE
fix use of jl_n_threads in gc; scale default collect interval

### PR DESCRIPTION
**Pull request by JeffBezanson**
~~`The first commit fixes a bug where `jl_gc_init` saw `jl_n_threads == 0` because it hadn't been set yet. The second commit scales the default collect interval by the number of threads. Otherwise, adding more threads basically translates directly to spending a higher % of runtime in GC.~~

This makes the collect interval fully thread-local, allowing it to naturally scale (if allocation happens on n threads, the effective interval is n times bigger). This also removes the dependence of gc_init on the number of threads (which wasn't working, since jl_n_threads wasn't initialized yet).

### CODE REVIEW

File: src/gc.c [hunk 1]
```diff
@@ -3029,6 +3029,9 @@ void jl_gc_init(void)
     arraylist_new(&finalizer_list_marked, 0);
     arraylist_new(&to_finalize, 0);
 
+    assert(jl_n_threads > 0);
+    default_collect_interval = default_collect_interval * jl_n_threads;
```
Comment (vchuravy) [inline src/gc.c]: Is linear scaling advisable? Now we have much longer interval in serial sections of the code, leading to generally more memory usage. Maybe some form of sublinear scaling would make sense.

Comment (JeffBezanson) [reply src/gc.c]: Just depends how much time you'd like to spend in GC :) So far I've found that anything less than this means more threads => greater % GC time and it's quite depressing when you're trying to get something to scale.

**Comment by chethega** [comment]
If I understand this issue right, the `collect_interval` describes the amount of memory slack (how much memory are we willing to waste on maybe-dead objects). Only one thread can run the gc, and typically all other cores idle in this time (unless we can feed them with something like BLAS). If we move up from e.g. 1 to 8 threads without changing the slack, then the amount and percentage of core-cycles spent on gc will remain the same, but these will represent a larger fraction of real-time. In effect, the cost of gc has gone up 8x, because we additionally burn all the possible work that the other 7 cores could have done in this time.

So you propose to simply linearly scale up the amount of slack by the number of cores; then the effective cost of gc stays the same. 

Did I understand this right?

This looks weird to me: If we can afford the larger slack, then we should use it regardless of the number of threads. If we cannot afford the larger slack, then a larger number of threads does not make it more affordable. 

In other words, maybe we want a sensible way for users to tell julia how much they value memory against real time against core-time. Especially since extra memory slack is often almost free until it suddenly becomes almost unaffordable (if the system has N gig of ram and is not doing anything else, then there is not a lot of downside to using up this memory; but we cannot use more memory than we have, and at some threshold the kernel will start doing stuff like swapping or evicting stuff from very useful caches).

**Comment by vtjnash** [comment]
I think there’s a couple mitigating factors. When the threads are in use, that also implies that other processes on the system aren’t using that time/memory. But more significantly, reclaiming space makes allocation much cheaper, but gc scales with the amount of memory still in use. With more threads, it may take more work to reach the same GC operational ratio.

**Comment by JeffBezanson** [comment]
Yes good points. I'm going to redo this; I think increasing the interval itself is not quite right, since that also causes code that uses only one thread (out of many available) to use more memory, which shouldn't be strictly necessary. I think the thing to do is really simple: just make the allocation counters thread-local, and GC as soon as any one thread hits the interval. That way single-thread code behaves the same as before, but if more threads are running *and allocating* then the effective interval naturally scales with the number of threads.

File: src/gc.c [hunk 2]
```diff
@@ -2913,8 +2892,15 @@ JL_DLLEXPORT void jl_gc_collect(int full)
 {
     jl_ptls_t ptls = jl_get_ptls_states();
     if (jl_gc_disable_counter) {
-        gc_num.deferred_alloc += (gc_num.allocd + gc_num.interval);
-        gc_num.allocd = -(int64_t)gc_num.interval;
+        size_t allocd = 0;
+        for (int i = 0; i < jl_n_threads; i++) {
+            jl_ptls_t ptls2 = jl_all_tls_states[i];
```
Comment (vtjnash) [inline src/gc.c]: you aren't sync'd here with the other threads--it's invalid to write through `jl_all_tls_states`

File: src/gc.c [hunk 3]
```diff
@@ -2622,8 +2602,7 @@ JL_DLLEXPORT int64_t jl_gc_total_bytes(void)
     jl_gc_num_t num = gc_num;
     combine_thread_gc_counts(&num);
     // Sync this logic with `base/util.jl:GC_Diff`
```
Comment (vtjnash) [inline src/gc.c]: todo?
